File: nng_comm/src/subscriber.cpp
#include "nng_comm/subscriber.hpp"
#include "nng_comm/message.hpp"
#include <iostream>

namespace nng_comm {

Subscriber::Subscriber() : initialized_(false), running_(false) {
    socket_.id = 0;
}

Subscriber::~Subscriber() {
    stop_async();
    close();
}

bool Subscriber::init(const std::string& address) {
    int rv;

    // Create subscriber socket
    if ((rv = nng_sub0_open(&socket_)) != 0) {
        std::cerr << "nng_sub0_open: " << nng_strerror(static_cast<nng_err>(rv)) << std::endl;
        return false;
    }

    // Subscribe to all topics (empty string)
    if ((rv = nng_sub0_socket_subscribe(socket_, "", 0)) != 0) {
        std::cerr << "nng_sub0_socket_subscribe: " << nng_strerror(static_cast<nng_err>(rv)) << std::endl;
        nng_socket_close(socket_);
        return false;
    }

    // Dial to publisher with NNG_FLAG_NONBLOCK for auto-reconnect
    // This allows the subscriber to be created before the publisher is up
    if ((rv = nng_dial(socket_, address.c_str(), nullptr, NNG_FLAG_NONBLOCK)) != 0) {
        std::cerr << "nng_dial: " << nng_strerror(static_cast<nng_err>(rv)) << std::endl;
        nng_socket_close(socket_);
        return false;
    }

    address_ = address;
    initialized_ = true;
    return true;
}

bool Subscriber::receive(nlohmann::json& msg, int timeout_ms) {
    if (!initialized_) {
        std::cerr << "Subscriber not initialized" << std::endl;
        return false;
    }

    // Set receive timeout
    if (timeout_ms >= 0) {
        nng_duration timeout = timeout_ms;
        nng_socket_set_ms(socket_, NNG_OPT_RECVTIMEO, timeout);
    }

    // Receive message using message API
    nng_msg* msgp = nullptr;
    int rv = nng_recvmsg(socket_, &msgp, 0);

    if (rv != 0) {
        if (rv != NNG_ETIMEDOUT) {
            std::cerr << "nng_recvmsg: " << nng_strerror(static_cast<nng_err>(rv)) << std::endl;
        }
        return false;
    }

    // Deserialize
    try {
        void* buf = nng_msg_body(msgp);
        size_t size = nng_msg_len(msgp);
        msg = Message::deserialize(static_cast<uint8_t*>(buf), size);
        nng_msg_free(msgp);
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Deserialize error: " << e.what() << std::endl;
        nng_msg_free(msgp);
        return false;
    }
}

void Subscriber::set_callback(std::function<void(const nlohmann::json&)> callback) {
    callback_ = callback;
}

void Subscriber::start_async() {
    if (!initialized_ || !callback_) {
        std::cerr << "Cannot start async: not initialized or no callback set" << std::endl;
        return;
    }

    running_ = true;
    recv_thread_ = std::thread(&Subscriber::receive_loop, this);
}

void Subscriber::stop_async() {
    if (running_) {
        running_ = false;
        if (recv_thread_.joinable()) {
            recv_thread_.join();
        }
    }
}

void Subscriber::receive_loop() {
    while (running_) {
        nlohmann::json msg;
        if (receive(msg, 100)) {  // 100ms timeout
            if (callback_) {
                callback_(msg);
            }
        }
    }
}

void Subscriber::close() {
    if (initialized_) {
        nng_socket_close(socket_);
        initialized_ = false;
    }
}

} // namespace nng_comm

File: nng_comm/scripts/subscriber.py
import threading
from typing import Callable, Optional

import pynng
from .message import Message


class Subscriber:
    def __init__(self):
        self._socket = None
        self._initialized = False
        self._callback = None
        self._running = False
        self._thread = None

    def init(self, address: str) -> bool:
        try:
            self._socket = pynng.Sub0()
            self._socket.subscribe(b"")
            self._socket.dial(address, block=False)
            self._initialized = True
            return True
        except Exception as e:
            print(f"Subscriber init failed: {e}")
            self._initialized = False
            return False

    def receive(self, timeout_ms: int = -1) -> Optional[dict]:
        if not self._initialized:
            return None
        try:
            if timeout_ms >= 0:
                self._socket.recv_timeout = timeout_ms
            else:
                self._socket.recv_timeout = None
            data = self._socket.recv()
            return Message.deserialize(data)
        except pynng.Timeout:
            return None
        except Exception as e:
            print(f"Subscriber receive failed: {e}")
            return None

    def set_callback(self, callback: Callable[[dict], None]):
        self._callback = callback

    def start_async(self):
        if not self._initialized or self._running:
            return
        self._running = True
        self._thread = threading.Thread(target=self._receive_loop, daemon=True)
        self._thread.start()

    def _receive_loop(self):
        while self._running:
            try:
                self._socket.recv_timeout = 100
                data = self._socket.recv()
                if self._callback:
                    self._callback(Message.deserialize(data))
            except pynng.Timeout:
                continue
            except Exception:
                if self._running:
                    continue
                break

    def stop_async(self):
        self._running = False
        if self._thread is not None:
            self._thread.join()
            self._thread = None

    def close(self):
        self.stop_async()
        if self._socket is not None:
            self._socket.close()
            self._socket = None
            self._initialized = False

    def is_initialized(self) -> bool:
        return self._initialized

    def __enter__(self):
        return self

    def __exit__(self, *args):
        self.close()
